wasm/validation.cpp

```diff
@@ -0,0 +1,520 @@
+    return is_valid(mt, 1ULL << 16);
+}
+
+// https://webassembly.github.io/spec/core/valid/types.html#match-limits
+// https://webassembly.github.io/spec/core/valid/types.html#memories
+constexpr bool is_match(Limits const &l1, Limits const &l2) {
+    if (l1.min >= l2.min) {
+        if (!l2.max.has_value()) {
+            return true;
+        }
+
+        if (l1.max.has_value() && l1.max <= l2.max) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// https://webassembly.github.io/spec/core/valid/types.html#functions
+constexpr bool is_match(FunctionType const &f1, FunctionType const &f2) {
+    return f1 == f2;
+}
+
+// https://webassembly.github.io/spec/core/valid/types.html#tables
+constexpr bool is_match(TableType const &t1, TableType const &t2) {
+    return is_match(t1.limits, t2.limits) && t1.element_type == t2.element_type;
+}
+
+// https://webassembly.github.io/spec/core/valid/types.html#globals
+constexpr bool is_match(GlobalType const &g1, GlobalType const &g2) {
+    return g1 == g2;
+}
+
+// https://webassembly.github.io/spec/core/appendix/algorithm.html#validation-algorithm
+struct ControlFrame {
+    Instruction i;
+    std::vector<ValueType> params;
+    std::vector<ValueType> results;
+    std::size_t stack_height = 0;
+    bool unreachable = false;
+};
+
+struct Validator {
+    std::vector<ValueOrUnknown> value_stack; // operand stack
+    std::vector<ControlFrame> control_stack;
+
+    void push_val(ValueOrUnknown const &);
+    tl::expected<ValueOrUnknown, ValidationError> pop_val();
+    tl::expected<ValueOrUnknown, ValidationError> pop_val_expect(ValueOrUnknown const &);
+    void push_vals(std::vector<ValueType> const &);
+    tl::expected<std::vector<ValueOrUnknown>, ValidationError> pop_vals(std::vector<ValueType> const &);
+    void push_ctrl(Instruction const &, std::vector<ValueType> const &, std::vector<ValueType> const &);
+    tl::expected<ControlFrame, ValidationError> pop_ctrl();
+    std::vector<ValueType> const &label_types(ControlFrame &);
+    void unreachable();
+};
+
+void Validator::push_val(ValueOrUnknown const &val) {
+    value_stack.push_back(val);
+}
+
+tl::expected<ValueOrUnknown, ValidationError> Validator::pop_val() {
+    assert(!control_stack.empty());
+
+    if (value_stack.size() == control_stack.back().stack_height && control_stack.back().unreachable) {
+        return Unknown{};
+    }
+
+    if (value_stack.size() == control_stack.back().stack_height) {
+        return tl::unexpected{ValidationError::ValueStackUnderflow};
+    }
+
+    auto val = value_stack.back();
+
+    value_stack.pop_back();
+
+    return val;
+}
+
+tl::expected<ValueOrUnknown, ValidationError> Validator::pop_val_expect(ValueOrUnknown const &expected) {
+    auto actual = pop_val();
+
+    if (!actual.has_value()) {
+        return tl::unexpected{actual.error()};
+    }
+
+    if (actual != expected && !std::holds_alternative<Unknown>(*actual) && !std::holds_alternative<Unknown>(expected)) {
```

Comment:
Maybe a separate `if (std::holds_alternative<Unknown>(*actual) || std::holds_alternative<Unknown>(expected)) { return actual; }` before this would be clearer?

Reply:
I think it reads better the way it's written now. "If the popped value wasn't what you expected, and neither the actual or expected values are Unknown, error".